src/v/iceberg/schema_json.cc

```diff
@@ -0,0 +1,80 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "iceberg/schema_json.h"
+
+#include "iceberg/datatypes.h"
+#include "iceberg/datatypes_json.h"
+#include "iceberg/json_utils.h"
+#include "iceberg/schema.h"
+#include "json/document.h"
+
+namespace iceberg {
+
+schema parse_schema(const json::Value& v) {
+    auto type = parse_required_str(v, "type");
+    if (type != "struct") {
+        throw std::invalid_argument(
+          fmt::format("Schema has type '{}' instead of 'struct'", type));
+    }
+    int32_t schema_id = parse_required_i32(v, "schema-id");
+    auto identifier_fids_json = parse_optional(v, "identifier-field-ids");
+    chunked_hash_set<nested_field::id_t> identifier_fids;
+    if (identifier_fids_json.has_value()) {
+        if (!identifier_fids_json->get().IsArray()) {
+            throw std::invalid_argument(
+              fmt::format("Schema has type '{}' instead of 'struct'", type));
+        }
+        for (const auto& id_json : identifier_fids_json->get().GetArray()) {
+            if (!id_json.IsInt()) {
+                throw std::invalid_argument(fmt::format(
+                  "Schema has non-int 'identifier-field-ids' value: {}",
+                  id_json.GetType()));
+            }
+            identifier_fids.emplace(id_json.GetInt());
+        }
+    }
+    auto struct_type = parse_struct(v);
+    return schema{
+      .sstruct = std::move(struct_type),
+      .schema_id = schema::id_t{schema_id},
+      .identifier_field_ids = std::move(identifier_fids),
+    };
+}
+
+} // namespace iceberg
+
+namespace json {
+
+void rjson_serialize(
+  json::Writer<json::StringBuffer>& w, const iceberg::schema& s) {
```

Comment:
This looks like it works, but is it necessary? It looks like schema.sstruct holds the entire original schema, including schema-id, type, fields, and identifier-field-ids. Would it work to just serialize schema.sstruct?

Reply:
You're right that they're very similar, but yes this is necessary because schema_struct only includes the type specification of the schema, rather than any additional metadata associated with the schema. Note how struct_type does not include a schema-id, or an identifier-field-ids member.

From https://iceberg.apache.org/spec/#schemas

>Schemas are serialized as a JSON object with the same fields as a struct in the table below, and the following additional fields...